packages/blob-storage/indexer/indexer.go

```
@@ -0,0 +1,388 @@
+		Password: cfg.DBPassword,
+		Database: cfg.DBDatabase,
+	})
+	if err != nil {
+		return err
+	}
+
+	i.ethClient = client
+	i.beaconURL = cfg.BeaconURL
+	i.taikoL1 = taikoL1
+	i.startHeight = cfg.StartingBlockID
+	i.db = db
+	i.wg = &sync.WaitGroup{}
+	i.ctx = ctx
+	i.cfg = cfg
+
+	return nil
+}
+
+func (i *Indexer) Start() error {
+	i.wg.Add(1)
+
+	var processingBlockHeight uint64
+
+	if i.startHeight != nil {
+		processingBlockHeight = *i.startHeight
+	} else {
+		n, err := i.getLatestBlockID()
+		if err != nil {
+			return err
+		}
+
+		processingBlockHeight = n
+	}
+
+	// get the latest header
+	header, err := i.ethClient.HeaderByNumber(i.ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	// the end block is the latest header.
+	endBlockID := header.Number.Uint64()
+
+	var defaultBlockBatchSize uint64 = 50
+
+	slog.Info("fetching batch block events",
+		"startHeight", processingBlockHeight,
+		"endblock", endBlockID,
+		"batchsize", defaultBlockBatchSize,
+	)
+
+	for j := processingBlockHeight; j < endBlockID; j += defaultBlockBatchSize {
+		end := processingBlockHeight + uint64(defaultBlockBatchSize)
+		// if the end of the batch is greater than the latest block number, set end
+		// to the latest block number
+		if end > endBlockID {
+			end = endBlockID
+		}
+
+		// filter exclusive of the end block.
+		// we use "end" as the next starting point of the batch, and
+		// process up to end - 1 for this batch.
+		filterEnd := end - 1
+
+		slog.Info("block batch", "start", j, "end", filterEnd)
+
+		opts := &bind.FilterOpts{
+			Start:   processingBlockHeight,
+			End:     &filterEnd,
+			Context: i.ctx,
+		}
+
+		group, groupCtx := errgroup.WithContext(i.ctx)
+
+		events, err := i.taikoL1.FilterBlockProposed(opts, nil, nil)
+		if err != nil {
+			return err
+		}
+
+		for events.Next() {
+			event := events.Event
+			group.Go(func() error {
+				if err := i.storeBlob(groupCtx, event); err != nil {
+					slog.Error("error storing blob", "error", err)
```

Shall we return the error / retry here? since we want to save every txlist blob into database?

---

retry and error handled